interp/interp.go

```diff
@@ -146,18 +150,142 @@ func (pc PiecewiseConstant) Predict(x float64) float64 {
 	if i < 0 {
 		return pc.ys[0]
 	}
-	// i < len(pci.xs)
 	if x == pc.xs[i] {
 		return pc.ys[i]
 	}
 	n := len(pc.xs)
 	if i == n-1 {
-		// x > pci.xs[i]
 		return pc.ys[n-1]
 	}
 	return pc.ys[i+1]
 }
 
+// PiecewiseCubic is a left-continuous, piecewise cubic 1-dimensional interpolator.
+type PiecewiseCubic struct {
+	// Interpolated X values.
+	xs []float64
+
+	// Coefficients of interpolating cubic polynomials, with
+	// len(xs) - 1 rows and 4 columns. The interpolated value
+	// for xs[i] <= x < xs[i + 1] is defined as
+	//   sum_{k = 0}^3 coeffs.At(i, k) * (x - xs[i])^k
+	// To guarantee left-continuity, coeffs.At(i, 0) == ys[i].
+	coeffs *mat.Dense
+
+	// Last interpolated Y value, corresponding to xs[len(xs) - 1].
+	lastY float64
+}
+
+// Predict returns the interpolation value at x.
+func (pc PiecewiseCubic) Predict(x float64) float64 {
+	i := findSegment(pc.xs, x)
+	if i < 0 {
+		return pc.coeffs.At(0, 0)
+	}
+	m := len(pc.xs) - 1
+	if x == pc.xs[i] {
+		if i < m {
+			return pc.coeffs.At(i, 0)
+		}
+		return pc.lastY
+	}
+	if i == m {
+		return pc.lastY
+	}
+	dx := x - pc.xs[i]
+	a := pc.coeffs.RawRowView(i)
+	return ((a[3]*dx+a[2])*dx+a[1])*dx + a[0]
+}
+
+// FitWithDerivatives fits a piecewise cubic predictor to (X, Y, dY/dX) value
+// triples provided as three slices.
+// It returns an error if len(xs) < 2, elements of xs are not strictly
+// increasing, len(xs) != len(ys) or len(xs) != len(dydxs).
+func (pc *PiecewiseCubic) FitWithDerivatives(xs, ys, dydxs []float64) error {
+	n := len(xs)
+	if len(ys) != n {
+		return errors.New(differentLengths)
+	}
+	if len(dydxs) != n {
+		return errors.New(differentLengths)
+	}
+	if n < 2 {
+		return errors.New(tooFewPoints)
+	}
+	m := n - 1
+	pc.coeffs = mat.NewDense(m, 4, nil)
+	for i := 0; i < m; i++ {
+		dx := xs[i+1] - xs[i]
+		if dx <= 0 {
+			return errors.New(xsNotStrictlyIncreasing)
+		}
+		dy := ys[i+1] - ys[i]
+		// a_0
+		pc.coeffs.Set(i, 0, ys[i])
+		// a_1
+		pc.coeffs.Set(i, 1, dydxs[i])
+		// Solve a linear equation system for a_2 and a_3.
+		pc.coeffs.Set(i, 2, (3*dy-(2*dydxs[i]+dydxs[i+1])*dx)/dx/dx)
+		pc.coeffs.Set(i, 3, (-2*dy+(dydxs[i]+dydxs[i+1])*dx)/dx/dx/dx)
+	}
+	pc.xs = make([]float64, n)
+	copy(pc.xs, xs)
+	pc.lastY = ys[m]
+	return nil
+}
+
+// AkimaSplines is a left-continuous, piecewise cubic 1-dimensional interpolator
+// which can be fitted to (X, Y) value pairs without providing derivatives.
+// See https://www.iue.tuwien.ac.at/phd/rottinger/node60.html for more details.
+type AkimaSplines struct {
+	PiecewiseCubic
```

Comment:
The `AkimaSpline` type should not expose the whole `PiecewiseCubic` interface. The essence of this spline is how it chooses the derivatives so you don't want them to be freely customizable via `FitWithDerivatives`.

Reply:
OK. I won't export `fitWithDerivatives` for now.